room_access_rules/__init__.py

```diff
@@ -336,7 +336,7 @@ async def check_event_allowed(
         self,
         event: EventBase,
         state_events: StateMap[EventBase],
-    ) -> Union[bool, dict]:
```

Comment:
Can we pin down the dictionary's key and value types?

Reply:
The "Returns" section of the docstring needs updating to reflect this.

Reply:
We don't seem to ever return a dictionary here; I'm guessing that other modules can choose to do so?

Reply:
Actually, I find this pretty confusing. `Tuple[bool, Optional[dict]]` expresses four states:

- `True, None`
- `False, None`
- `True, dict()`
- `False, dict()`

but the docstrings seems to only describe three cases.

Reply:
Note that this is a specific implementation of https://matrix-org.github.io/synapse/latest/modules/third_party_rules_callbacks.html#check_event_allowed, which has more generic documentation on what modules should return.

> We don't seem to ever return a dictionary here; I'm guessing that other modules can choose to do so?

Yep (one module ran by a customer actually does that).



> Actually, I find this pretty confusing. `Tuple[bool, Optional[dict]]` expresses four states:
> 
>   * `True, None`
>   * `False, None`
>   * `True, dict()`
>   * `False, dict()`
> 
> but the docstrings seems to only describe three cases.

Yes, as the documentation for this callback implies, the second return value is ignored if the first one is `False`.

Reply:
> Can we pin down the dictionary's key and value types?

I'll update it to `Dict[str, Any]` - since this is meant to be an event dict I don't think there's a better way to do it (unless we have existing art for doing otherwise).

Reply:
> Note that this is a specific implementation of https://matrix-org.github.io/synapse/latest/modules/third_party_rules_callbacks.html#check_event_allowed, which has more generic documentation on what modules should return.

Aha, thank you, that is very enlightening.

Reply:
> > Can we pin down the dictionary's key and value types?
> 
> I'll update it to `Dict[str, Any]` - since this is meant to be an event dict I don't think there's a better way to do it (unless we have existing art for doing otherwise).

Done in https://github.com/vector-im/synapse-room-access-rules/pull/3/commits/ed6918cd087d25d239f61f405fa2fc97d146b992